### ISSUE TITLE
Probability prediction for multiclass with L1/L2 loss

### ISSUE THREAD

**Issue by jemus42** [PlantedML/randomPlantedForest#42]
Since L1 and L2 losses are not guaranteed to return valid probability predictions, i.e. predictions can exceed the [0, 1] interval, we have to figure out how to handle that in the multiclass setting.

In the binary case, our previous approach of [truncating class probabilities to [0, 1]](https://github.com/PlantedML/randomPlantedForest/blob/87607c2ad7a0a0ab1f4d3559a406cf4ff69091d0/R/predict_rpf.R#L128-L132) was sufficient, since we only make predictions `p` for the positive class and define the prediction for the negative class as `1 - p`, ensuring a sum of 1.

In the multiclass setting, this truncation still applies, so individual class predictions for `k` classes don't fall out of the plausibility range, but the sum is not.  

## What do we do?

Instead of truncating, we could divide predicted class probabilities by the sum of the predictions, normalizing to a maximum of 1.
However, this would skew the individual class predictions in one way or another, and if we implement something like this on the R side then the C++ implementation and the R wrapper would yield different results - so I'd argue this needs to be addressed at a low level.

## Example cases for testing/reference:

``` r
library(randomPlantedForest)
xdf <- na.omit(palmerpenguins::penguins)
rpfit <- randomPlantedForest::rpf(species ~ ., data = xdf, loss = "L1")
pred <- predict(rpfit, xdf, type = "prob")
pred$sum <- rowSums(pred)

# Implausible cases are returned as-is:
pred[pred$sum > 1, ]
#> # A tibble: 205 × 4
#>    .pred_Adelie .pred_Chinstrap .pred_Gentoo   sum
#>           <dbl>           <dbl>        <dbl> <dbl>
#>  1        1            0.00970        0       1.01
#>  2        1            0.00321        0       1.00
#>  3        0.954        0.0715         0       1.03
#>  4        0.996        0.0283         0       1.02
#>  5        0.995        0.0419         0       1.04
#>  6        1            0.00931        0       1.01
#>  7        0.964        0.0455         0       1.01
#>  8        0.976        0.0604         0       1.04
#>  9        1            0.000516       0       1.00
#> 10        0.574        0.396          0.0502  1.02
#> # … with 195 more rows
```

And with `mlr3`, we don't get any predictions at all (I'd argue `predict.rpf()` probably should also at least warn):

``` r
library(mlr3)
library(mlr3extralearners) # for rpf learner
# Penguins, dropping missing obs
test_task <- tsk("penguins")
ids <- complete.cases(test_task$data())
test_task$filter(which(ids))

lrn_rpf <- lrn("classif.rpf", loss = "L1", predict_type = "prob")
lrn_rpf$train(test_task)

# mlr3 correctly refuses to accept sum of class probabilities exceed 1:
pred <- lrn_rpf$predict(test_task)
#> Error: Probabilities for observation 1 do sum up to 1.012296 != 1

# So we don't get any predictions at all:
pred
#> Error in eval(expr, envir, enclos): object 'pred' not found
```

<sup>Created on 2022-11-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by jemus42** [comment]
Addendum: I don't think truncating or even normalizing is going to help.

For the "raw" L1 predictions I'm getting not a single class prediction in the [0, 1] range:

``` r
library(randomPlantedForest)
# real data, train/test split
xdf <- na.omit(palmerpenguins::penguins)
train_ids <- sample(333, 222)
train <- xdf[train_ids, ]
test <- xdf[setdiff(1:333, train_ids), ]

# fit
rpfit <- randomPlantedForest::rpf(species ~ ., data = train, loss = "L1")

# needs preprocessing due do factor variables etc.
train_mat <- randomPlantedForest:::preprocess_predictors_predict(rpfit, test[ , -1])

# Raw predictions from C++
pred_raw <- rpfit$fit$predict_matrix(train_mat, 0)

pred_raw[1:10, ]
#>           [,1]        [,2]       [,3]
#>  [1,] 1.354417 -0.17641083 -0.1780065
#>  [2,] 1.354417 -0.17641083 -0.1780065
#>  [3,] 1.218130 -0.04711969 -0.1710106
#>  [4,] 1.218130 -0.04711969 -0.1710106
#>  [5,] 1.218130 -0.04711969 -0.1710106
#>  [6,] 1.218130 -0.04711969 -0.1710106
#>  [7,] 1.354417 -0.17641083 -0.1780065
#>  [8,] 1.218130 -0.04711969 -0.1710106
#>  [9,] 1.218130 -0.04711969 -0.1710106
#> [10,] 1.354417 -0.17641083 -0.1780065

# All predicted values are either greater 1 or smaller 0
all(pred_raw[, 1] > 1)
#> [1] TRUE
all(pred_raw[, 2] < 0)
#> [1] TRUE
all(pred_raw[, 3] < 0)
#> [1] TRUE
```

<sup>Created on 2022-11-08 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So we either need to apply `softmax()` after all and I misunderstood something earlier in that regard, or something's fishy in the implementation.

Also note that in `predict.rpf()` we have `type = "prob"` which _needs_ to yield valid probabilities, summing to 1 and all, but we also have `type = "numeric"` which is allowed to take any value and may be interesting for other reasons (component-wise plotting? I'm not sure).

**Comment by jemus42** [comment]
Fixed in [bc009f2](https://github.com/PlantedML/randomPlantedForest/commit/bc009f2c782d67c50a3364c3f631a6241df33f6a), we now just divide by the sum of predictions in this case.

Note that `predict(..., type = "numeric")` will still return the "raw", unmodified predictions.
